ADR/0043-artifact-constructors.md

```diff
@@ -0,0 +1,104 @@
+
+Date: 2024-10-21
+
+## Status
+
+Proposed
+
+## Authors
+
+- Krunoslav Pavic
+- Andrew McNamara
+
+## Context
+
+This ADR aims to cover the proposed functionality for creating additional Artifacts as part of the Konflux build/test/release process.
+These artifacts would vary in type and would cover use cases where either the single built image or entire sets of images
+from the user's application are needed to generate one or more artifacts.
+
+Some limitations with our current approach that can be mitigated with this ADR:
+* The ability to produce multiple artifacts with a single PipelineRun
+* Reduced need for in-PR nudging (one of the common use cases for this is the operator bundle builds as mentioned below)
+
+
+## Use Cases in Detail
+
+* A team produces operands and operators. When integration service creates a Snapshot, it uses a constructor that
+  actually builds a bundle. When releasing, the team releases that single bundle artifact. That constructor also has an extract
+  operation to list all related images that are referenced, and that extraction generates the list for release service to operate on
+* A team produces a set of images with ko. The build task generates/pushes all images with ko and then produces a "build artifact"
+  OCI manifest which includes an immutable reference to each of the ko-built artifacts. When integration service creates a Snapshot,
+  it uses a constructor that doesn't need to create a new artifact. When releasing, the team releases that single artifact.
+  The constructor also has an extract operation to list all the ko-built artifacts and that extraction generates the list for release service to operate on.
+* A team produces a set of images that are just a group of artifacts that should be released together.
+  The artifacts are built via separate PipelineRuns. When integration service creates a Snapshot, it uses a constructor
+  that saves the Snapshot CR as an OCI blob. When releasing, the team releases that single artifact.
+  The constructor also has an extract operation to list all artifacts that were in the Snapshot and that extraction
+  generates the list for release service to operate on.
+
+
+## Decision
+
+A new API will be introduced which instructs the Integration service to create a Tekton pipeline which will generate the
+artifact.
+Once built, the artifact reference will be added to the Snapshot's metadata so it can be discovered during testing or release process.
+Additionally, the integration service will wait for the artifact constructor to finish before starting the integration testing pipelines
+in order to enable users to run tests using the new artifacts.
+
+### The ArtifactConstructor API
+
+The proposal here is to introduce the new ArtifactConstructor API which will define Tekton pipelines that will generate individual artifacts.
+These pipelines would be executed after a Snapshot is generated and before the integration test pipelines start.
```

Comment:
I'm again totally confused. Above you say `Once built, the artifact reference will be added to the Snapshot's metadata` and now you say `API which will define Tekton pipelines that will generate individual artifacts. These pipelines would be executed after a Snapshot is generated`.

Reply:
Yeah, we may need a diagram here for the proposed workflow:

1. The Snapshot gets created (in response to a build pipeline or manually etc.)
2. The Artifact generation pipelineRuns are started and generate artifacts based on the information in the Snapshot
3. References to the generated artifacts will be added to the Snapshot's metadata (annotations) to provide a central place where we can find all generated artifacts and testing information.